src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -68,7 +68,7 @@ impl<'a, 'tcx> Visitor<'tcx> for BorrowckCtxt<'a, 'tcx> {
     }
 
     fn visit_fn(&mut self, fk: FnKind<'tcx>, fd: &'tcx hir::FnDecl,
-                b: hir::ExprId, s: Span, id: ast::NodeId) {
+                b: hir::BodyId, s: Span, id: ast::NodeId) {
```

Comment:
Why shouldn't it? That's equivalent to `({*&Foo(22)}).0` which is perfectly fine AFAIK.
The problem would be if `X` where `static` (and only until we get miri).